### PR TITLE
Propagate distributed tracing

### PR DESCRIPTION
**Pull request by labbati**
### Description

This pool request add distributed tracing headers propagation for guzzle and curl integration

### Readiness checklist
- [x] [Changelog entry](docs/changelog.md) added, if necessary
- [x] Tests added for this feature/bug

### CODE REVIEW

File: src/DDTrace/Integrations/Curl/CurlIntegration.php [hunk 1]
```diff
@@ -50,5 +57,52 @@ public static function load()
...
+        if (!Configuration::instance()->isDistributedTracingEnabled()) {
+            return;
+        }
+
+        $currentHttpHeaders = ArrayKVStore::getForResource($ch, 'http_headers', []);
+        if (is_array($currentHttpHeaders)) {
+            $tracer = GlobalTracer::get();
+            $context = $tracer->getActiveSpan()->getContext();
+            $carrier = Headers::colonSeparatedValuesToHeadersMap($currentHttpHeaders);
+            $tracer->inject($context, HTTP_HEADERS, $carrier);
```
[the remark this answers is not in the document]
Comment (labbati) [reply src/DDTrace/Integrations/Curl/CurlIntegration.php]: I think I may not follow here.
The problem I had is that different libraries handle headers differently. So for example, in the case of duplicate headers you may receive `['headers1: value1, value2']` for curl and `['header1' => ['value1', 'value2']]` in Guzzle. So would an additional propagator resolve this problem? :(

File: src/DDTrace/Integrations/Curl/CurlIntegration.php [hunk 2]
```diff
@@ -50,5 +57,52 @@ public static function load()
...
+
+            curl_setopt($ch, CURLOPT_HTTPHEADER, Headers::headersMapToColonSeparatedValues($carrier));
```
[the remark this answers is not in the document]
Comment (labbati) [reply src/DDTrace/Integrations/Curl/CurlIntegration.php]: this is a good idea, let me see how I can change the implementation.

[the remark this answers is not in the document]
Comment (labbati) [reply src/DDTrace/Integrations/Curl/CurlIntegration.php]: Ahhh now I remember, it has to be in the exec method to have the proper span activated (for parent span id).

File: src/DDTrace/Integrations/Curl/CurlIntegration.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (pawelchcki) [reply src/DDTrace/Integrations/Curl/CurlIntegration.php]: Then possible two additional propagators might be in order. 
My point is however that in case of curl we can expect values like:
`[['header' => 'value1"], ['header' => 'value2']]` so converting the parameters first to associative array
means that we will lose `['header' => 'value']` when converting back array of tuples

[the remark this answers is not in the document]
Comment (labbati) [reply src/DDTrace/Integrations/Curl/CurlIntegration.php]: > My point is however that in case of curl we can expect values like
Ah ok I did not know this. Looking at them it looked like I only could get them in the form `['header: value1, value2']`, but let me check.

File: src/DDTrace/Integrations/Curl/CurlIntegration.php [hunk 3]
```diff
@@ -50,5 +55,56 @@ public static function load()
...
+    {
+        if (!Configuration::instance()->isDistributedTracingEnabled()) {
+            return;
+        }
+
+        $httpHeaders = ArrayKVStore::getForResource($ch, Formats\CURL_HTTP_HEADERS, []);
+        if (is_array($httpHeaders)) {
+            $tracer = GlobalTracer::get();
+            $context = $tracer->getActiveSpan()->getContext();
+            $tracer->inject($context, Formats\CURL_HTTP_HEADERS, $httpHeaders);
```
Comment (pawelchcki) [inline src/DDTrace/Integrations/Curl/CurlIntegration.php]: ❤️  😄

File: src/DDTrace/Util/ArrayKVStore.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace DDTrace\Util;
+
+/**
+ * A key value store that stores metadata into an array. If you have an object that you can use as a carrier, then
+ * prefer ObjectKVStore as is provides a better performance. Use this if you do not have an object you can use as
+ * a carrier.
+ */
+class ArrayKVStore
```
Comment (pawelchcki) [inline src/DDTrace/Util/ArrayKVStore.php]: 👍

File: src/DDTrace/Util/ArrayKVStore.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    private static function getResourceKey($resource)
+    {
+        // Converting to integer a resource results in the "unique resource number assigned to the resource by PHP at
+        // runtime":
+        //   - http://php.net/manual/en/language.types.integer.php#language.types.integer.casting
+        // Resource ids are guaranteed to be unique per script execution:
+        //   - http://www.php.net/manual/en/language.types.string.php#language.types.string.casting
+        return intval($resource);
+    }
+}
```
Comment (pawelchcki) [inline src/DDTrace/Util/ArrayKVStore.php]: This class is 💯

File: tests/Unit/Util/ArrayKVStoreTest.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+final class ArrayKVStoreTest extends TestCase
+{
+    private $resource1;
+    private $resource2;
+
+    protected function setUp()
+    {
+        parent::setUp();
+        $this->resource1 = fopen('php://memory', 'r');
```
Comment (pawelchcki) [inline tests/Unit/Util/ArrayKVStoreTest.php]: TIL :D

File: tests/Integration/Frameworks/Laravel/Version_4_2/composer.json [hunk 7]
```diff
@@ -25,6 +25,7 @@
 			"DDTrace\\": "../../../../../src/DDTrace/"
 		},
 		"files": [
+			"../../../../../src/DDTrace/Formats.php",
```
Comment (pawelchcki) [inline tests/Integration/Frameworks/Laravel/Version_4_2/composer.json]: 😢 would be great to find some approach - not requiring this.
But this is out of scope of this PR.

Comment (labbati) [reply tests/Integration/Frameworks/Laravel/Version_4_2/composer.json]: I know :(
